tilespec.ts

```diff
@@ -1,54 +1,213 @@
+            bottom = Math.round(workArea.origin.y + positiveTileNumber * elemSize.height);
+        } else if (this.rdc.types.y == 'approx_percentage') {
+            const snappedToGrid = Math.round(this.gridHeight * this.rdc.xy.y);
+            bottom = Math.round(workArea.origin.y + snappedToGrid * elemSize.height);
+        } else {
+            bottom = Math.round(workArea.origin.y + workArea.size.height * this.rdc.xy.y);
+        }
+
         return new Rect(
-            new XY(
-                workArea.origin.x + this.luc.x * elemSize.width,
-                workArea.origin.y + this.luc.y * elemSize.height),
-            new Size((this.rdc.x + 1 - this.luc.x) * elemSize.width,
-                (this.rdc.y + 1 - this.luc.y) * elemSize.height));
+            new XY(left, top),
+            new Size(right - left - 1, bottom - top - 1)
+        );
     }
 
     get gridSize(): GridSize {
         return new GridSize(this.gridWidth, this.gridHeight);
     }
 
     viewSize(): GridSize {
-        const sizeXY = this.rdc.minus(this.luc);
+        const sizeXY = this.rdc.xy.minus(this.luc.xy);
         return new GridSize(sizeXY.x + 1, sizeXY.y + 1);
     }
 
     isFullscreen(): boolean {
         return this.viewSize().equals(this.gridSize);
     }
+
+    /**
+     * Converts negative coordinates (e.g. -1:-1) to a positive format on a specified grid.
+     * If x or y is a positive number, it is ignored.
+     * E.g. -1:-1 on a 3:3 grid is a 3:3, as well as -1:3.
+     */
+    _convertNegativeCoord(gridEdges: number, coord: number) {
+        if (coord < 0) {
+            return gridEdges + coord + 1;
+        } else {
+            return coord;
+        }
+    }
+}
+
+export const MAX_TUPLE_MEMBER_VALUE = Number.MAX_SAFE_INTEGER;
+
+/**
+ * Tuple Holder represents a single starting or ending point (x and y coordinates),
+ * as well as the type of the coordinate - "tile", "approx_percentage" or "percentage" now.
+ *
+ * E.g. ~0.75:0.75 is {X:0.75,Y:0.75}, types - 'percentage' & 'percentage'
+ * approximate - true.
+ */
+export class TupleHolder {
+    raw: string;
+    xy: XY;
+    types: CoordinateTypesHolder;
+
+    constructor(raw: string) {
+        this.raw = raw;
+
+        const gssk = this.raw.split(':');
+
+        this._validateTuple(gssk);
+
+        this.xy = this._parseTuple(gssk);
+        this.types = this._parseTypes(gssk);
+    }
+
+    toString() {
+        return this.raw;
+    }
+
+    _parseTuple(tuple: Array<string>) {
+        const x = this._parseCoordinate(tuple[0]);
+        const y = this._parseCoordinate(tuple[1]);
+
+        return new XY(x, y);
+    }
+
+    _parseTypes(tuple: Array<string>) {
+        const typeX = this._parseType(tuple[0]);
+        const typeY = this._parseType(tuple[1]);
+
+        return new CoordinateTypesHolder(typeX, typeY);
+    }
+
+    _parseCoordinate(coord: string) {
+        return Number(coord.replace('~', ''));
+    }
+
+    _parseType(coord: string) {
```

Comment:
Don't use _ for private instance variables, use the private keyword of Typescript: https://google.github.io/styleguide/tsguide.html#naming-style